[Element 1.4] Implement new API browser.scrollBy() and browser.scrollTo()

honglax opened the pull request:
Implementing new API for Element.

**browser.scrollBy(x: number, y: number, [,option])**
```
- x: How many pixels to scroll by, along the x-axis (horizontal). Positive values will scroll to the right, while negative values will scroll to the left. 
- x can take window.innerWidth as a special value.
- y: How many pixels to scroll by, along the y-axis (vertical). Positive values will scroll down, while negative values scroll up.
- y can take window.innerHeight as a special value.
- option: is an object containing ScrollToOptions.

E.g.: 
await browser.scrollBy(100, 0, { behavior: 'smooth' })
await browser.scrollBy('window.innerWidth', 0, { behavior: 'smooth' })
await browser.scrollBy(0, 'window.innerHeight', { behavior: 'smooth' })

const innerHeight: number = await browser.evaluate(() => window.innerHeight)
const innerWidth: number = await browser.evaluate(() => window.innerWidth)
await browser.scrollBy(innerWidth, innerHeight, { behavior: 'smooth' })
```


**browser.scrollTo(target: Point, [,option])**
```
- point is an array of x (number) and y (number) co-ordinate
- options is an object containing ScrollToOptions.

E.g.: await browser.scrollTo([69, 96], { behavior: 'smooth', block: 'nearest' })
```


**browser.scrollTo(target: ElementHandle, [,option])**
```
E.g.: 
const title = await browser.findElement(By.css('#title'))
await browser.scrollTo(title, { behavior: 'smooth', block: 'center', inline: 'center' })
```


**browser.scrollTo(target: Locator, [,option])**
`E.g.: await browser.scrollTo(By.visibleText('submit'), { behavior: 'smooth' })`


**browser.scrollTo(ScrollDirection: string, [,option])**
```
ScrollDirection can be one of the following: top, bottom, left, or right 
E.g.: await browser.scrollTo("top", { behavior: 'smooth' })
```

phileba commented:
`browser.scrollTo([69, 96], { behavior: "smooth" })` @sonnguyenmau @honglax do you think we need to support point array, it will scroll point by point in this array, maybe we should have an option like `scrollDelay` 🤔

sonnguyenmau commented:
Hi a @phileba , I don't think we need to support this. The user can always use 2 separate scrolls to achieve the same result. I think we should support 2 special params, which are window.innerHeight and window.innerWidth. What do you think?

phileba commented:
> Hi a @phileba , I don't think we need to support this. The user can always use 2 separate scrolls to achieve the same result. I think we should support 2 special params, which are window.innerHeight and window.innerWidth. What do you think?

okay Sơn, so how to support this?

sonnguyenmau commented:
instead of using only number, users can input window.innerHeight or window.innerWidth as special values. Don't worry a Phi, @honglax will handle this

phileba commented:
> instead of using only number, users can input window.innerHeight or window.innerWidth as special values. Don't worry a Phi, @honglax will handle this

okay Sơn, I believe him :v

HieuLeKMS commented on packages/core/src/page/types.ts:
```diff
@@ -202,3 +202,6 @@ export interface TargetLocator {
 	frame(id: number | string | ElementHandle): Promise<void>
 }
+
+// Using these type below for API scrollTo()
```
```suggestion
// Using the type below for API scrollTo()
```

honglax replied:
Resolved.

HieuLeKMS commented on packages/core/src/runtime/Browser.ts:
```diff
@@ -600,4 +601,131 @@ export class Browser<T> implements BrowserInterface {
+		const behavior = scrollOptions?.behavior ?? 'auto'
+
+		if (!this.isCorrectScrollBehavior(behavior)) {
+			throw new Error('The input behavior is not correct (Must be "auto" or "smooth").')
```
It's not really matter but `throw new Error(...)` & `throw Error(...)` are equivalent

honglax replied:
So should we use `throw Error(...)` instead of `throw new Error(...)`?

HieuLeKMS replied:
I think either is fine. I just informed you in case you didn't know.

honglax replied:
Got it. Thanks for your advice.

phileba replied:
please be careful when you throw an Error like this, it will stop the test when running on Flood.

phileba commented on packages/core/src/runtime/Browser.ts:
```diff
@@ -600,4 +601,143 @@ export class Browser<T> implements BrowserInterface {
+
+	// Check the target has type Locator or not
+	private isLocator(target: Locator | ElementHandle | Point | ScrollDirection): target is Locator {
+		return !Array.isArray(target) && (target as Locator).find !== undefined
```
This check is not safe I think, it will be an issue if others target (e.g: ElementHandle) has method `find`

honglax replied:
Resolved.

phileba commented on packages/core/src/runtime/Browser.ts:
```diff
@@ -600,4 +601,143 @@ export class Browser<T> implements BrowserInterface {
+	private isElementHandle(
+		target: Locator | ElementHandle | Point | ScrollDirection,
+	): target is ElementHandle {
+		return (target as ElementHandle).element !== undefined
```
the same as above, I think we should have a method like `type` for Locator and ElementHandle, then it will return exactly which type of the target.

honglax replied:
So should I add a method `type` for **Locator** and **ElementHandle**? Is it a breaking change or not?

phileba replied:
I don't want to change the interface so find another solution with a smaller scope

honglax replied:
Got it.

HieuLeKMS replied:
You might want to try the operator `instanceof`

phileba replied:
yep, so we can distinguish `Locator` and `ElementHandle` by:
```ts
import { BaseLocator } from '../page/Locator'
import { ElementHandle as TargetElementHandle } from '../page/Locators'
...
if (target instanceof BaseLocator){
// locator
} else if (target instanceof TargetElementHandle) {
// element handle
}

honglax replied:
Unfortunately, `Locator` and `ElementHandle` just are `Interface` hence we can not use `instanceof` with them.

phileba replied:
Yep I know but I'm talking about class `BaseLocator` and `ElementHandle` :slap:

HieuLeKMS replied:
You can import and check with the classes that implement those interfaces as @phileba suggested

honglax replied:
Got it.

honglax replied:
Resolved.

phileba commented on packages/core/src/runtime/Browser.ts:
```diff
@@ -600,4 +601,143 @@ export class Browser<T> implements BrowserInterface {
+
+	private isPoint(target: Locator | ElementHandle | Point | ScrollDirection): target is Point {
+		return (
+			Array.isArray(target) &&
+			target.length === 2 &&
+			typeof target[0] === 'number' &&
+			typeof target[1] === 'number'
```
is it enough?
```suggestion
			Array.isArray(target) &&
			typeof target[0] === 'number' &&
			typeof target[1] === 'number'
```

honglax replied:
To prevent the end-user using `[1, 2, 'hello world']` or something like this.

phileba replied:
in this case, the target is Point, so it should be `[number, number]`, the compile will fail if the input is `[1, 2, 'hello world']` 🤔

honglax replied:
I want to check the length because I want to throw a friendly error message to the end-user if they input the incorrect type of param. The compile will fail but maybe the end-user will not get a friendly error message.

phileba replied:
okay

honglax replied:
Resolved.

phileba commented on packages/core/src/runtime/Browser.ts:
```diff
@@ -600,4 +601,143 @@ export class Browser<T> implements BrowserInterface {
+	}
+
+	private isCorrectScrollBehavior(behavior: string): behavior is ScrollBehavior {
+		return behavior === 'auto' || behavior === 'smooth'
```
```suggestion
		return ['auto', 'smooth'].includes(behavior)
```

honglax replied:
Resolved.

phileba commented on packages/core/src/runtime/Browser.ts:
```diff
@@ -600,4 +601,143 @@ export class Browser<T> implements BrowserInterface {
+		])
+
+		if (this.isPoint(target)) {
+			// target is Point
```
the function already told this target is a Point so we don't need this comment I think

honglax replied:
Resolved

phileba commented on packages/core/src/runtime/Browser.ts:
```diff
@@ -600,4 +601,143 @@ export class Browser<T> implements BrowserInterface {
+
+		if (this.isPoint(target)) {
+			// target is Point
+			top = target[1]
+			left = target[0]
```
```suggestion
			[left, top] = target
```

phileba commented on packages/core/src/runtime/Browser.ts:
```diff
@@ -600,4 +601,143 @@ export class Browser<T> implements BrowserInterface {
+					)
+			}
+		} else if (this.isLocator(target) || this.isElementHandle(target)) {
+			// target is Locator or ElementHandle
```
🔥

honglax replied:
Resolved.

phileba commented on packages/core/src/runtime/Browser.ts:
```diff
@@ -600,4 +601,143 @@ export class Browser<T> implements BrowserInterface {
+			}
+		} else if (this.isLocator(target) || this.isElementHandle(target)) {
+			// target is Locator or ElementHandle
+			const targetEl = await this.findElement(target)
+			await targetEl.element.evaluate(
+				(el, scrollOptions: ScrollIntoViewOptions) => {
+					el.scrollIntoView(scrollOptions)
+				},
+				{ behavior, block, inline },
+			)
+			return
```
for this block, we return when complete the action so I think we should move this to the top as an early return pattern.

honglax replied:
Resolved.

phileba commented on examples/flood-element-challenge.ts:
```diff
@@ -0,0 +1,427 @@
+		for (let x = 1; x <= numOfPage; x++) {
+			const productCards = await browser.findElements(By.xpath('//div[@data-test-card]'))
+			for (let y = 0; y < productCards.length; y++) {
+				// await browser.scrollTo(productCards[y])
+				// await browser.evaluate(y => {
+				// 	const el = document.querySelectorAll('div[data-test-card]')[y]
+				// 	el.scrollIntoView({ block: 'nearest' })
+				// }, y)
```
🔥

honglax replied:
Resolved.

phileba commented on examples/flood-element-challenge.ts:
```diff
@@ -0,0 +1,427 @@
+				await browser.mouse.down()
+				await browser.mouse.up()
+
+				// Or you can click into the close button in the modal
+				// const closeModalBtnEl = await browser.findElement(By.css('button[data-test-product-detail-modal-close]'))
+				// await closeModalBtnEl.click()
```
🔥

honglax replied:
Resolved.

phileba commented on packages/core/src/runtime/Browser.ts:
```diff
@@ -600,4 +603,141 @@ export class Browser<T> implements BrowserInterface {
+		])
+
+		if (this.isLocator(target) || this.isElementHandle(target)) {
+			const targetEl = await this.findElement(target)
```
if target already is an `ElementHandle`, do we need to `findElement` again?

honglax replied:
`ElementHandle.findElement()` returns an ElementHandle hence I use it to avoid using too much if-else.

phileba replied:
It will call `Browser.findElement`, then it will try to cast the `ElementHandle` to `Locator` in `locatableToLocator` function, then call `locator.find` to return an `ElementHandle`. Luckily the ElementHandle itself already has the method `find` too so it works but the logic is incorrect and we might have an issue with this in the future 🤔

honglax replied:
So I'll use `findElement()` if the **target Element** is `Locator` to prevent any issue.

phileba replied:
I think so, it's easy to maintain later

honglax replied:
Thanks for your advice.

HieuLeKMS commented on packages/core/src/runtime/Browser.spec.ts:
```diff
@@ -228,4 +237,124 @@ describe('Browser', () => {
+		})
+		test('scrollTo', async () => {
+			await browser.scrollTo('bottom', { behavior: 'smooth' })
+			await browser.wait(2)
```
What are these `browser.wait(second)` for? I think these will increase a good amount of time for running unit tests. Should we remove them?

honglax replied:
When we use `scroll` with `behavior`, it will take time to scroll to the target position. Hence, I need to wait to make sure the scroll action has finished.

honglax commented:
Hi @HieuLeKMS, I've refactored the unit-test file to remove the redundant line of code. I've also added a HTML file to using for test the API scrollTo() and scrollBy().

HieuLeKMS commented:
LGTM. 🚢 when you're 😃
